Move `django_ssl.conf` into `django.conf` (single HTTP/HTTPS server)

Raruto opened the pull request:
**List of changes:**

- delete `config/_nginx/django_ssl.conf`
- update `config/nginx/django.conf` accordingly in order to use a single HTTP/HTTPS server
- expose `WEBGIS_PUBLIC_HOSTNAME` environment variable to `nginx` container (`docker-compose.yml`)
- expose `WEBGIS_PUBLIC_HOSTNAME` environment variable to `nginx` container (`docker-compose-consumer.yml`)
- update `README.md` section related to `HTTPS additional setup`
- refactor `run_certbot.sh` (to improve readability and terminal output)

Closes: https://github.com/g3w-suite/g3w-suite-docker/issues/67, closes: https://github.com/g3w-suite/g3w-suite-docker/issues/68

Raruto committed:
Move `django_ssl.conf` into `django.conf`

- delete `config/_nginx/django_ssl.conf`
- update `config/nginx/django.conf` in order to use a singl HTTP/HTTPS server
- expose `WEBGIS_PUBLIC_HOSTNAME`
 environment variable to `nginx` container (`docker-compose.yml` and `docker-compose-consumer.yml`)
- update `README.md` section related to `HTTPS additional setup`

Raruto commented on config/nginx/django.conf:
```diff
@@ -36,36 +33,43 @@ server {
-  server_name dev.g3wsuite.it;
+  # if ($scheme = http) {
+  #     return 301 https://$server_name$request_uri;
+  # }
```
Check if this is stricly necessary:

```nginx
if ($scheme = http) {
  return 301 https://$server_name$request_uri;
}
```

or if the following rule could be enough:

```nginx
listen 443 default ssl;
```

**NB:**

- [NGINX > If is Evil…](https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/)
- [NGINX > Pitfalls and Common Mistakes](https://www.nginx.com/resources/wiki/start/topics/tutorials/config_pitfalls/)

wlorenzetti replied:
The if statement I think is necessary, otherwise we would have a 'loop' of constant redirections. With if we have redirection only on http protocol.

Raruto commented on config/nginx/django.conf:
```diff
@@ -36,36 +33,43 @@ server {
+  listen 8080;
+
   # !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
   # Uncomment the following lines if you want activate https
```
Check if we can add here an environment variable instead of keeping the whole code commented, eg:

```nginx
if ($G3W_ENABLE_SSL) {

  listen 443 default;

  ...

}
```

wlorenzetti replied:
For use environment variables into nginx configurations files at today is necessary use third part plugin.

Raruto replied:
@wlorenzetti from what I read [here](https://github.com/docker-library/docs/tree/master/nginx#using-environment-variables-in-nginx-configuration-new-in-119) it looks like that's feature already included in the docker version (BTW at the moment I don't have access to a real server and I can't tell how much of this changes are correct..)

wlorenzetti replied:
@Raruto I just try for example to add new env variable: `$NGINX_CLIENT_MAX_BODY_SIZE`
I add it to `django.conf`:

```nginx
client_max_body_size $NGINX_CLIENT_MAX_BODY_SIZE;
```
and to `docker-compose.yml` file into `nginx` service:

```yml
...
    environment:
      - WEBGIS_PUBLIC_HOSTNAME
      - NGINX_CLIENT_MAX_BODY_SIZE
...      
```
But it not works, into nginx container log I found:
```
nginx: [emerg] "client_max_body_size" directive invalid value in /etc/nginx/conf.d/django.conf:1
```

As you can read [here](https://github.com/docker-library/docs/tree/master/nginx#using-environment-variables-in-nginx-configuration-new-in-119) we have to create a nginx.conf.template file with environment variables, i.e. ${NGINX_CLIENT_MAX_BODY_SIZE} so that _envsubst_ at startup can create the .conf file correctly.

Raruto replied:
@wlorenzetti just a quick test to see how it could works: https://github.com/g3w-suite/g3w-suite-docker/pull/77/commits/8b3e51cee29e2ee5a9bfd45a391b4bcace606d88 (ref. to the `NGINX_CLIENT_MAX_BODY_SIZE` envsubst)

Probably, It won't work out of the box moving from g3w-suite v3.4 to v3.5 (having to edit the docker-compose.yml file as well), anyway, let me know if you see any other major issues.

For future reference, this is the code used by the docker image to make that substitution: [docker-nginx/entrypoint/20-envsubst-on-templates.sh](https://github.com/nginxinc/docker-nginx/blob/fef51235521d1cdf8b05d8cb1378a526d2abf421/entrypoint/20-envsubst-on-templates.sh)

wlorenzetti replied:
@Raruto I just pushed a commit to fix your changes: https://github.com/g3w-suite/g3w-suite-docker/pull/77/commits/36e6142b202e39c5d1881830e9e32f07bef29b79.

With your modifications `envsubst` builds the configuration file for nginx but removes the 'suffix', which in our case is` .conf` as seen from the entrypoint https://github.com/nginxinc/docker-nginx/blob/fef51235521d1cdf8b05d8cb1378a526d2abf421/entrypoint/20-envsubst-on-templates.sh#L28.

It was enough to rename `django.conf` to `django.conf.template` and change `NGINX_ENVSUBST_TEMPLATE_SUFFIX` to `.template`

Raruto commented on README.md:
```diff
@@ -94,15 +94,14 @@ the conf file will be mounted into docker container at runtime to PGSERVICEFILE
 
 To active https with LetsEncrypt just follow the following instructions:
```
Currently I am unable to verify the correctness of these statements.

wlorenzetti replied:
@Raruto I just check and it works fine.

Raruto committed:
Merge branch 'dev' into nginx

Raruto committed:
add sample `NGINX_ENVSUBST_TEMPLATE_SUFFIX`

Commit:
Fix for nginx `envsubst`.

Raruto committed:
refactor `run_certbot.sh`

- add some variables to improve readability
- echo some minimal info to terminal (current step)
- remove `docker pull certbot/certbot` step within README.md

Raruto commented on README.md:
```diff
@@ -97,15 +97,13 @@ the conf file will be mounted into docker container at runtime to PGSERVICEFILE
+- update `WEBGIS_PUBLIC_HOSTNAME` environment variable within the `.env` file
 - launch `./run_certbot.sh`
-- activate 301 redirect into `config/nginx/django.conf`
 - restart compose
```
ambiguous, rewrite better to indicate the exact terminal command needed to restart compose

wlorenzetti replied:
So is it clear? https://github.com/g3w-suite/g3w-suite-docker/pull/77/commits/0c30c8c5cc06a846481bfeae54f3c5955a81a35b

Raruto replied:
> terminal command needed to restart compose
>
> `docker compose down && docker compose up -d`

by chance, does this do the same? `docker compose up -d --force-recreate`

Ref: [Compose UP options](https://docs.docker.com/engine/reference/commandline/compose_up/#options)

Raruto commented on README.md:
```diff
@@ -94,15 +94,14 @@ the conf file will be mounted into docker container at runtime to PGSERVICEFILE
-- activate 301 redirect into `config/nginx/django.conf`
 - restart compose
 - make sure the certs are renewed by adding a cron job with `crontab -e` and add the following line:
   `0 3 * * * /<path_to_your_docker_files>/run_certbot.sh`
```
So why do we need to define a certbot container in `docker-compose.yml` as well?

https://github.com/g3w-suite/g3w-suite-docker/blob/e5695a13af0a0a3a9c8802945027e7bafdc035b6/docker-compose.yml#L95-L100


https://github.com/g3w-suite/g3w-suite-docker/blob/e5695a13af0a0a3a9c8802945027e7bafdc035b6/run_certbot.sh#L24-L31

wlorenzetti replied:
@Raruto Right, I'll do some tests

wlorenzetti replied:
Remove certbot image from docker compose file https://github.com/g3w-suite/g3w-suite-docker/pull/77/commits/9cfbbed1d69d9cbca6bdceeadce03b84aacec04a

wlorenzetti replied:
I just check ona ne clear VM with new domain and it works fine

Commit:
Bash fix variable declaration.

wlorenzetti commented on run_certbot.sh:
```diff
@@ -16,11 +16,11 @@ if [ "${WEBGIS_DOCKER_SHARED_VOLUME}" = "" ]; then
     exit 1
 fi
 
-certs_folder      = "${WEBGIS_DOCKER_SHARED_VOLUME}/certs/letsencrypt"
```
@Raruto in bash there must be no spaces in the declaration of the variables.

Raruto replied:
> @Raruto in bash there must be no spaces in the declaration of the variables.

👍

Commit:
Remove certbot image from docker compose files.

wlorenzetti committed:
Merge branch 'dev' into nginx

Commit:
Add line to close https activation section.

Commit:
Update https section into README.md

Commit:
Update contributors.

wlorenzetti commented:
@Raruto I checked every point and tested, if you don't want propose new changes for me it possible merge it on dev.

Raruto commented:
> The if statement I think is necessary, otherwise we would have a 'loop' of constant redirections. With if we have redirection only on http protocol.

@wlorenzetti given the declarative nature of nginx confing it is not immediately clear whether the ssl options are also sent to the browser when requesting the page as `http` version:

```nginx
  ssl_certificate /etc/letsencrypt/live/$WEBGIS_PUBLIC_HOSTNAME/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/$WEBGIS_PUBLIC_HOSTNAME/privkey.pem;

  include /etc/letsencrypt/options-ssl-nginx.conf;
  ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

  resolver 8.8.8.8;

  # written like this its doubtful if this condition is executed before or after the previous ssl options
  if ($scheme = http) {
      return 301 https://$server_name$request_uri;
  }
```

Probably an optimal solution (also in terms of future customizations) could be to keep two separate server sections and share the common portions through the `include` directive:

```nginx
server {
    listen 80;
    server_name example.org;
    
    include mime_types;
    include default_routes;
    
    location / {
        return 301 https://$host$request_uri;
    }    
}

server {
    listen 443 ssl;
    server_name example.org;
    
    include mime_types;
    include default_routes;
    
    include /etc/letsencrypt/options-ssl-nginx.conf;

    ....
    
    location / {
        proxy_pass http://example.org; #for demo purposes
    }
}
```

> The if statement I think is necessary, otherwise we would have a 'loop' of constant redirections. With if we have redirection only on http protocol.

In https://github.com/g3w-suite/g3w-suite-docker/pull/77#discussion_r979963606 i was thinking if it could be feasible to replace the if condition with the `default_server` directive (see also: [How nginx processes a request](https://nginx.org/en/docs/http/request_processing.html)), anyway, as you can see I still don't have a very clear idea about it..

BTW, another solution, which would not require changes to the nginx configuration files, could be to always activate ports 80 and 443 inside the container and `expose` them to the public only when needed (e.g. via `docker-compose.override.yml`, ref: https://github.com/g3w-suite/g3w-suite-docker/pull/79)

So if there is no rush, I would wait a moment longer before merging this.
